src/librustc_target/spec/apple_base.rs

```diff
@@ -51,3 +51,63 @@ pub fn macos_llvm_target(arch: &str) -> String {
     let (major, minor) = macos_deployment_target();
     format!("{}-apple-macosx{}.{}.0", arch, major, minor)
 }
+
+#[cfg(target_os = "macos")]
+pub fn sysroot(sdk: &str) -> Result<Option<String>, String> {
+    // Like Clang, allow the `SDKROOT` environment variable used by Xcode to define the sysroot.
+    if let Some(sdk_root) = env::var("SDKROOT").ok() {
+        let actual_sdk_path = sdk_path(sdk)?;
+        let sdk_root_p = Path::new(&sdk_root);
+        // Ignore `SDKROOT` if it's not a valid path.
+        if !sdk_root_p.is_absolute() || sdk_root_p == Path::new("/") || !sdk_root_p.exists() {
+            return Ok(Some(actual_sdk_path));
+        }
+        // Ignore `SDKROOT` if it's clearly set for the wrong platform, which may occur when we're
+        // compiling a custom build script while targeting iOS for example.
+        return Ok(Some(match sdk {
+            "iphoneos" if sdk_root.contains("iPhoneSimulator.platform")
+                || sdk_root.contains("MacOSX.platform") => actual_sdk_path,
+            "iphonesimulator" if sdk_root.contains("iPhoneOS.platform")
+                || sdk_root.contains("MacOSX.platform") => actual_sdk_path,
+            "macosx" | "macosx10.15" if sdk_root.contains("iPhoneOS.platform")
+                || sdk_root.contains("iPhoneSimulator.platform") => actual_sdk_path,
+            _ => sdk_root,
+        }))
+    }
+    Ok(None)
+}
+
+// `xcrun` is only available on macOS.
+#[cfg(not(target_os = "macos"))]
+pub fn sysroot(_sdk: &str) -> Result<Option<String>, String> {
+    if let Some(sdk_root) = env::var("SDKROOT").ok() {
+        let sdk_root_p = Path::new(&sdk_root);
+        // Use `SDKROOT` only if it's a valid path.
+        if sdk_root_p.is_absolute() && sdk_root_p != Path::new("/") && sdk_root_p.exists() {
+            return Ok(Some(sdk_root));
+        }
+    }
+    Ok(None)
+}
+
+#[cfg(target_os = "macos")]
+fn sdk_path(sdk_name: &str) -> Result<String, String> {
+    let res = std::process::Command::new("xcrun")
+        .arg("--show-sdk-path")
+        .arg("-sdk")
+        .arg(sdk_name)
+        .output()
+        .and_then(|output| {
+            if output.status.success() {
+                Ok(String::from_utf8(output.stdout).unwrap())
+            } else {
+                let error = String::from_utf8(output.stderr);
+                let error = format!("process exit with error: {}", error.unwrap());
```

Comment:
I see that this was just copied from before, but could this `.unwrap` and the one above be avoided by using `from_utf8_lossy` instead?